File: NWMv4/scripts/apply_donor_constraints.R
# apply a few constraints to donors 
apply_donor_constraints <- function(rec, donors, dists, pars, dtAttr) {
  
  # 1. narrow down to donors with the same snowiness category
  snowy <- subset(dtAttr,id==rec & tag=="receiver")$snowy
  snowy1 <- subset(dtAttr,id%in%donors & tag=="donor")$snowy
  ix1 <- which(snowy1==snowy)
  if (length(ix1)>0) {dists <- dists[ix1]; donors <- donors[ix1]}
  
  # 2. further narrow down to those donors within maximum spatial distance defined
  ix1 <- which(dists <= pars$maxSpaDist)
  if (length(ix1)>0) {dists <- dists[ix1]; donors <- donors[ix1]}
  
  # 3. further narrow down based on screening attributes
  for (att1 in names(pars$maxAttrDiff)) {
    ix1 <- which(abs(subset(dtAttr,id==rec & tag=="receiver")[[att1]] - 
                       subset(dtAttr,id%in%donors & tag=="donor")[[att1]])<= pars$maxAttrDiff[[att1]])
    if (length(ix1)>0) {dists <- dists[ix1]; donors <- donors[ix1]}
  }
  
  # 4. further narrow down to donors in the same HSG
  hsg <- subset(dtAttr,id==rec & tag=="receiver")$hsg
  hsg1 <- subset(dtAttr,id%in%donors & tag=="donor")$hsg
  ix1 <- which(hsg1==hsg)
  if (length(ix1)>0) {dists <- dists[ix1]; donors <- donors[ix1]}
  
  return(list(donor=donors, dist=dists))
  
} #function apply_donor_constraints

File: NWMv4/algorithm/apply_donor_constraints.py
# apply a few constraints to donors identified (e.g., via Gower's distance or another technique)
def func(rec, donors, dists, pars, dtAttr):
  
    import numpy as np
  
    # 1. narrow down to donors with the same snowiness category
    snowy = dtAttr.query("id==@rec & tag=='receiver'")['snowy']
    snowy1 = dtAttr.query("id in @donors & tag=='donor'")['snowy']
    ix1 = snowy1.isin(snowy)
    if sum(ix1) > 0:
        dists = np.array(dists)[ix1]
        donors = np.array(donors)[ix1]
  
    # 2. further narrow down to those donors within maximum spatial distance defined
    ix1 = dists <= pars['general']['maxSpaDist']
    if sum(ix1) > 0:
        dists = np.array(dists)[ix1]
        donors = np.array(donors)[ix1]  
  
    # 3. further narrow down based on screening attributes
    for att1 in pars['general']['maxAttrDiff'].keys():
        ix1 = abs(np.array(dtAttr.query("id==@rec & tag=='receiver'")[att1]) - \
            np.array(dtAttr.query("id in @donors & tag=='donor'")[att1])) \
                <= pars['general']['maxAttrDiff'][att1]
        if sum(ix1) > 0:
            dists = np.array(dists)[ix1]
            donors = np.array(donors)[ix1]  
        
    # 4. further narrow down to donors in the same HSG
    hsg = dtAttr.query("id==@rec & tag=='receiver'")['hsg']
    hsg1 = dtAttr.query("id in @donors & tag=='donor'")['hsg']
    ix1 = hsg1.isin(hsg)
    if sum(ix1) > 0:
        dists = np.array(dists)[ix1]
        donors = np.array(donors)[ix1]
  
    return donors, dists
  